### PR TITLE
C# IR: More support for `ForStmt`s

### PR DESCRIPTION
**Pull request by AndreiDiaconu1**
This PR adds support for multiple initializers and updates in a `ForStmt`.
Also contains some autoformating.

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedCall.qll [hunk 1]
```diff
@@ -10,7 +10,7 @@ private import semmle.code.csharp.ir.internal.IRCSharpLanguage as Language
 
 /**
  * The IR translation of a call to a function. The function can be a normal function
- * (ie. `MethodCall`) or a constructor call (ie. `ObjectCreation`). Notice that the
+ * (eg. `MethodCall`) or a constructor call (eg. `ObjectCreation`). Notice that the
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedCall.qll]: Why this change? Also, in general, I believe it is `i.e.` and `e.g.`, and not `ie.` and `eg.`.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 2]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
-  TranslatedExpr getUpdate() { result = getTranslatedExpr(stmt.getAnUpdate()) }
+  TranslatedExpr getUpdate(int index) { result = getTranslatedExpr(stmt.getUpdate(index)) }
 
   private predicate hasUpdate() { exists(stmt.getAnUpdate()) }
 
+  private int initializersNo() { result = count(stmt.getAnInitializer()) }
+
+  private int updatesNo() { result = count(stmt.getAnUpdate()) }
+
+  private predicate initializerIndices(int index) { index in [0 .. initializersNo() - 1] }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: I would rename to `initializerIndex` (singular).

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 3]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
+
+  private predicate updateIndices(int index) {
+    index in [initializersNo() .. initializersNo() + updatesNo() - 1]
+  }
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: I think I would rewrite to
```
  private int updateIndex(int index) {
    result in [0 .. updatesNo() - 1] and
    index = initializersNo() + result
  }
```
and then change
```
updateIndices(id) and result = this.getUpdate(id - initializersNo())
```
above to
```
result = this.getUpdate(updateIndex(id))
```

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 4]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
+    initializerIndices(id) and result = this.getDeclAndInit(id)
     or
-    id = 1 and result = this.getCondition()
+    updateIndices(id) and result = this.getUpdate(id - initializersNo())
     or
-    id = 2 and result = this.getUpdate()
+    id = initializersNo() + updatesNo() and result = this.getCondition()
     or
-    id = 3 and result = this.getBody()
+    id = initializersNo() + updatesNo() + 1 and result = this.getBody()
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: The indices of `getChild()` do not need to be consecutive, nor start from `0`, right(?) In case there are no initializers and updates, `getChild()` will only have the body at index `1`.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 5]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
+    then result = this.getDeclAndInit(0).getFirstInstruction()
     else result = this.getFirstConditionInstruction()
   }
 
   override Instruction getChildSuccessor(TranslatedElement child) {
-    child = this.getDeclAndInit() and
+    exists(int index |
+      this.hasInitialization() and
+      child = this.getDeclAndInit(index) and
+      index < initializersNo() - 1 and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: This line is not needed.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 6]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
     if this.hasInitialization()
-    then result = this.getDeclAndInit().getFirstInstruction()
+    then result = this.getDeclAndInit(0).getFirstInstruction()
     else result = this.getFirstConditionInstruction()
   }
 
   override Instruction getChildSuccessor(TranslatedElement child) {
-    child = this.getDeclAndInit() and
+    exists(int index |
+      this.hasInitialization() and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Not needed.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 7]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
       child = this.getBody() and
       if this.hasUpdate()
-      then result = this.getUpdate().getFirstInstruction()
+      then result = this.getUpdate(0).getFirstInstruction()
       else result = this.getFirstConditionInstruction()
     )
     or
-    child = this.getUpdate() and result = this.getFirstConditionInstruction()
+    exists(int index |
+      this.hasUpdate() and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Not needed.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 8]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
+      child = this.getUpdate(index) and
+      index < updatesNo() - 1 and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Not needed.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll [hunk 9]
```diff
@@ -651,43 +651,70 @@ class TranslatedForStmt extends TranslatedLoop {
...
+      result = this.getUpdate(index + 1).getFirstInstruction()
+    )
+    or
+    this.hasUpdate() and
```
Comment (hvitved) [inline csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Not needed.

File: csharp/ql/test/library-tests/ir/ir/stmts.cs [hunk 10]
```diff
@@ -68,9 +68,15 @@ public static void tryCatchFinally()
...
+        for (int i = 0, j = 10; i < j; i++, j--)
+        {   
+            x = x - 1; 
+        }
+        
+        int a = 0, b = 10;
+        for (; a < b; ) 
         {
-            x = x - 1;
+            a++;
```
Comment (hvitved) [inline csharp/ql/test/library-tests/ir/ir/stmts.cs]: I suggest adding a test for
```
for ( ; ; )
{
    ....
}
```
as well.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedCall.qll (hunk 1 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedCall.qll]: This was part of the PR that got merged, probably wasn't picked up in the diff. Anyways, modified it to `i.e.` and `e.g.`.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 4 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Yes, the indices can be arbitrary numbers, so having the body at index `1` is ok.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 2 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Fixed.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 3 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: That seems clearer, refactored that bit.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 5 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Removed (I had those lines just for clarity).

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 6 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Removed.

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 7 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Removed

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 8 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Removed

File: csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll (hunk 9 above)
Comment (AndreiDiaconu1) [reply csharp/ql/src/semmle/code/csharp/ir/implementation/raw/internal/TranslatedStmt.qll]: Removed

File: csharp/ql/test/library-tests/ir/ir/stmts.cs (hunk 10 above)
Comment (AndreiDiaconu1) [reply csharp/ql/test/library-tests/ir/ir/stmts.cs]: Added the test case, seems to produce correct code.
